core/base/polymorphic_object.hpp

```diff
@@ -0,0 +1,521 @@
+ *                          hierarchy, has to be a subclass of polymorphic
+ *                          object
+ *
+ * @see EnablePolymorphicObject for creating a concrete subclass of
+ *      PolymorphicObject.
+ */
+template <typename AbstactObject, typename PolymorphicBase = PolymorphicObject>
+class EnableAbstractPolymorphicObject : public PolymorphicBase {
+public:
+    using PolymorphicBase::PolymorphicBase;
+
+    std::unique_ptr<AbstactObject> create_default(
+        std::shared_ptr<const Executor> exec) const
+    {
+        return std::unique_ptr<AbstactObject>{static_cast<AbstactObject *>(
+            this->create_default_impl(std::move(exec)).release())};
+    }
+
+    std::unique_ptr<AbstactObject> create_default() const
+    {
+        return this->create_default(this->get_executor());
+    }
+
+    std::unique_ptr<AbstactObject> clone(
+        std::shared_ptr<const Executor> exec) const
+    {
+        auto new_op = this->create_default(exec);
+        new_op->copy_from(this);
+        return new_op;
+    }
+
+    std::unique_ptr<AbstactObject> clone() const
+    {
+        return this->clone(this->get_executor());
+    }
+
+    AbstactObject *copy_from(const PolymorphicObject *other)
+    {
+        return static_cast<AbstactObject *>(this->copy_from_impl(other));
+    }
+
+    AbstactObject *copy_from(std::unique_ptr<PolymorphicObject> other)
+    {
+        return static_cast<AbstactObject *>(
+            this->copy_from_impl(std::move(other)));
+    }
+
+    AbstactObject *clear()
+    {
+        return static_cast<AbstactObject *>(this->clear_impl());
+    }
+};
+
+
+/**
+ * This macro implements the `self()` method, which is a shortcut for
+ * `static_cast<_type *>(this)`.
+ *
+ * It also provides a constant version overload. It is often useful when
+ * implementing mixins which depend on the type of the affected object, in which
+ * case the type is set to the affected object (i.e. the CRTP parameter).
+ */
+#define GKO_ENABLE_SELF(_type)                                    \
+    _type *self() noexcept { return static_cast<_type *>(this); } \
+                                                                  \
+    const _type *self() const noexcept                            \
+    {                                                             \
+        return static_cast<const _type *>(this);                  \
+    }
+
+
+/**
+ * ConvertibleTo interface is used to mark that the implementer can be converted
+ * to the object of ResultType.
+ *
+ * This interface is used to enable conversions between polymorphic objects.
+ * To mark that an object of type `U` can be converted to an object of type `V`,
+ * `U` should implement ConvertibleTo<V>.
+ * Then, the implementation of `PolymorphicObject::copy_from` automatically
+ * generated by `EnablePolymorphicObject` mixin will use RTTI to figure out that
+ * `U` implements the interface and convert it using the convert_to / move_to
```

Comment:
Maybe surround these such as `convert_to`, in case underscores are a problem at some point of the documentation generation (also because it's code).

Reply:
I think that if we surround it in underscores we'll disable doxygen's reference generation. But it could be wrong right now, maybe it needs convert_to() / move_to() to figure out it's looking for a function.

We'll have to check these things once we have doxygen generation.

Reply:
Oh I see it's intended to generate references, I thought it was simple code so proposed using ` here. If I understand the automatic link generation documentation, it looks like it requires either `::` before the name or `()` after it. Eventually, prefixed with the class name. We'll have to check all of those when we do proper documentation.